test/test_buf_error.c

```diff
@@ -0,0 +1,286 @@
+#include "test.h"
+#include "test_utils.h"
+
+#define MAX_LOOPS 5
+
+/*
+ * Here there is a small collection of test cases for when inflate must return Z_BUF_ERROR.
+ * 1. test_0_avail_in - no progress can be made because avail_in is 0.
+ * 2. test_small_avail_out tests - no progress can be made because avail_out is not enough
+ * to finish at once.
+ * 3. test_scp - replicates an issue found using libnxz with scp.
+ */
+
+int test_0_avail_in() {
+
+	z_stream strm = {0};
+	Byte compr[10], uncompr[10];
+	unsigned int outlen = 4096;
+	int rc = 0, ret = 0;
+
+	if ((rc = inflateInit(&strm)) != Z_OK){
+		printf("*** failed: inflateInt returned %d\n", rc);
+		return TEST_ERROR;
+	}
+
+	strm.avail_out = outlen;
+	strm.next_out = uncompr;
+	strm.next_in = compr;
+	strm.avail_in = 0;
+
+	if ((rc = inflate(&strm, Z_NO_FLUSH)) != Z_BUF_ERROR) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflate returned %d\n", rc);
+	}
+
+	if ((rc = inflateEnd(&strm)) != Z_OK) {
+		ret = TEST_ERROR;
+		printf("*** failed: inflateEnd returned %d\n", rc);
+	}
+
+	return ret;
+}
+
+int test_small_avail_out(){
+
+	z_stream strm = {0};
+	Byte *src, *compr, *uncompr;
+	int srclen = 128;
+	unsigned int comprlen = srclen*2;
+	unsigned int uncomprlen = 3;
+	int rc = 0, ret = 0;
+
+	generate_random_data(128);
+	src = ran_data;
+
+	if ((compr = (Byte*)calloc(comprlen, sizeof(Byte))) == NULL) {
+		printf("*** failed: could not allocate memory");
+		return TEST_ERROR;
+	}
+
+	if ((uncompr = (Byte*)calloc(uncomprlen, sizeof(Byte))) == NULL) {
```

Comment:
@lucmaga remember to free `compr` and `uncompr` before the end of the function. The test is failing under valgrind right now.